Handle empty query bodies at parse time and remove EmptyQueryBuilder

cbuescher opened the pull request:
As stated in #17540 we currently support empty query bodies like the filter in

```
"constant_score" : {  "filter" : { } }
```

in the query DSL. How these empty query bodies are handled depends on the query they are contained in. Upstream they are usually either ignored, converted to match all or no documents or bubbled up further in the query tree. While parsing they are currently represented as EmptyQueryBuilders, which serve no other purpose as to signal callers that it is the result of parsing an empty query body. 
When not handled anywhere else, it needs to be checked for on the shard when building the lucene query. This is trappy, so this PR changes the parsing of compound queries. Instead of returning QueryBuilder, the core query parsing method QueryShardContext#parseInnerQueryBuilder() now return an Optional<QueryBuilder> which is empty in the case of queries with empty body. This has the advantage of forcing caller code to deal with this sooner or later. when encountering empty Optionals, compound query builders now have the choice to ignore them, pass them on or rewrite to a different query, depending on context. 

Closes #17541

cbuescher commented on core/src/main/java/org/elasticsearch/index/percolator/PercolatorFieldMapper.java:
```diff
@@ -201,7 +197,8 @@ static Query toQuery(QueryShardContext context, boolean mapUnmappedFieldsAsStrin
         try {
-            return context.parseInnerQueryBuilder();
+            // NORELEASE not really sure what to do here
+            return context.parseInnerQueryBuilder().get();
```
I'm nor sure if this should error or ignore if the query hasn't been resolved to something at his point. Maybe @martijnvg has an opinion?

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/QueryParser.java:
```diff
@@ -38,7 +39,7 @@
      * @return the new QueryBuilder
      */
-    QB fromXContent(QueryParseContext parseContext) throws IOException;
+    Optional<QB> fromXContent(QueryParseContext parseContext) throws IOException;
```
Needed to change this signature so some QueryBuilders (like Boosting, ConstantScore, HasChild...) have the change to deffer resolving an empty Optional to something (thats what they also used to do before by returning 'null' to their caller in parse() (on 2.0) and now in doToQuery().

cbuescher committed:
Handle empty query bodies at parse time and remove EmptyQueryBuilder

Currently we support empty query clauses like the filter in

"constant_score" : {  "filter" : { } }

How these clauses are handled depends on the surrounding query.
They later are either ignored, converted to match all or no documents or
passed up further in the query hierarchy. During parsing these claues are
currently represented as EmptyQueryBuilders. When not handled anywhere else,
these special cases need to be checked for on the shard when building the
lucene query.

This is trappy, so this PR changes the parsing of compound queries. Instead
of returning QueryBuilder, the core query parsing method
QueryShardContext#parseInnerQueryBuilder() now return an Optional which can
be empty in the case of empty query clauses. This has the advantage of forcing
callers to deal with this sooner or later. When encountering empty Optionals,
compound query builders now have the choice to ignore them, pass them on or
rewrite to a different query, depending on context.

colings86 commented:
This looks good to me but I wonder if we should get another person to look at it (maybe @martijnvg @jpountz or @javanna ?) to get their opinion on the approach. The reason for this is that this is quite a big change to the query parsing (touches every query) to deal with a case that we are going to remove in 6.0 anyway. So although this is an improvement for 5.x to deal with the `{}` we are probably going to reverse this change in 6.0 and not return Optional from the query parse method since in 6.0 the parse method will always need to return a QueryBuilder and will not be allow to return null or anything that represents `{}`.

jpountz commented:
Agreed: I would rather like a soon-to-be-deprecated feature to be a bit buggy than to have deep implications on our internal  APIs.

cbuescher commented:
@colings86 @jpountz thanks for taking a look. I would still like to argue for getting this in: we only introduced the EmptyQueryBuilder as a means to handle the empty clauses on master to avoid having to deal with null values during the query refactoring. This is not really a bug, but it is really just a crutch that we would have avoided if we had had Optional<T> when we introduced it (@javanna might agree here). Now that we can represent optional values differently, I think we should do that rather than introduce a query builder that can get missused. Even if we plan to remove it in 6.0, it will stick around for a shile, but I think Optional<QueryBuilder> is a much better choice with Java 8. I also think the changes to query builder method signatures are rather small in most cases (just wrapping the builder in an Optional), but we don't have to transport those objects across the wire anymore if we deal with empty clauses on the coordinating node already.

javanna commented on core/src/main/java/org/elasticsearch/index/IndexService.java:
```diff
@@ -606,7 +608,11 @@ private Query parse(AliasMetaData alias, QueryShardContext shardContext) {
+                ParsedQuery parsedFilter = null;
+                if (innerQueryBuilder.isPresent()) {
+                    parsedFilter = shardContext.toFilter(innerQueryBuilder.get());
+                }
```
can we return directly instead of assigning this variable? I would find it a bit more readable

cbuescher replied:
Not sure this works, according to the old logic we need to call `query()` on the ParsedQuery but might get an NPE if we don't check parsedFilter for null as well. I might be missing something though.

javanna replied:
I may be missing something, does that mean that shardContext.toFilter may return null as well?

cbuescher replied:
I checked a few and found that e.g. `CommonTermsQueryBuilder#doToQuery()` can return null if the analyzer for the query doesn't provide any token. In `QueryShardContext#toQuery()` we do an explicit null-check and emit a MatchNoDocsQuery instead of null Query, in `toFilter()` we return null. I wonder if we should also return MatchNoDocsQuery there instead, then I could change the code above wo omitt the check, wdyt?

javanna replied:
I think we should consider null a bug here. I don't see good reasons why doToQuery should return null. Best fix would be to fix that at the source instead of having null check in QueryShardContent#toQuery . What does null mean in CommonTermsQueryBuilder? it should be responsibility of each query to construct a proper non null lucene query I think.

cbuescher replied:
Okay, so I added the early return, assuming `toFilter()` should always return a query. In case of empty inner query I think this method should still return null as `IndexService#aliasFilter()` actually documents and handles those cases and wouldn't like to break existing expectations here without further inspection.
+1 on re-visiting shard contexts `toQuery()` and `toFilter()` potentially returning null, but I'd like to do that in a separate PR.

javanna replied:
ok on a separate PR, thanks!

javanna commented:
I had a look at this and I like it. I agree that we would have used `Optional` from the beginning if we could, rather than introducing `EmptyQueryBuilder` to the transport layer, which can also be used from the java api. With this solution we handle the empty query case as soon as possible while parsing, we make it more explicit and that makes sure we don't have null lucene queries generated down the road. This is a good improvement and even if we remove this "feature" in 6.x this code will live for the whole 5.x series and make it better.

> Agreed: I would rather like a soon-to-be-deprecated feature to be a bit buggy than to have deep implications on our internal APIs.

@jpountz by deep implication you mean the fact that we use `Optional` in `fromXContent`, which will go away in 6.x ? I find it better than having `EmptyQueryBuilder` that gets serialized over the transport, which generates null lucene queries down the road and that is not explicit at all, so one has to remember to handle the null case. Optional solves it sooner and makes it more explicit. I think it's worth to get this in, but I may be missing some of the downsides. What do you think?

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -325,21 +325,17 @@ public static BoolQueryBuilder fromXContent(QueryParseContext parseContext) thro
                 case MUST:
-                    query = parseContext.parseInnerQueryBuilder();
-                    mustClauses.add(query);
+                    parseContext.parseInnerQueryBuilder().ifPresent(mustClauses::add);
```
along with these changes on some of the compound queries, i would like to see the corresponding change in toQuery where we don't have to handle the null lucene query case anymore, as only EmptyQueryBuilder could generate a null lucene query. Also we should test that toQuery never ever returns null in AbstractQueryTestCase.

cbuescher committed:
Removing handling of null lucene query where we catch this at parse time

cbuescher commented:
@javanna thanks for the review, I rebased and added a commit that removes the handling of null queries in the `toQuery` part where applicable. I'd still like to get this into 5.0 so we don't have to introduce EmptyQueryBuilder, even if we remove empty clauses altogether later. That is, if @colings86 and @jpountz don't have strong objections against it.

jpountz commented:
I am fine with this pull request either if we merge it to 5.x after master becomes 6.x, or if we make sure to move the API back to its current state once the branch is cut.

cbuescher commented:
@javanna I left one comment about possivle null return values in `QueryShardContext#toFilter()` and adressed your previous comments in the last commit. Mind to take another look?

cbuescher commented:
@javanna I changed the thing in IndexService#parse(), also opened https://github.com/elastic/elasticsearch/pull/18656 to remove the potential `null`return value in CommonTermsQueryBuilder.

cbuescher committed:
Add tests to check that toQuery() doesn't return null

javanna commented on core/src/main/java/org/elasticsearch/cluster/metadata/AliasValidator.java:
```diff
@@ -141,7 +142,10 @@ public void validateAliasFilter(String alias, byte[] filter, QueryShardContext q
+        Optional<QueryBuilder> parseInnerQueryBuilder = queryParseContext.parseInnerQueryBuilder();
+        if (parseInnerQueryBuilder.isPresent()) {
+            QueryBuilder queryBuilder = QueryBuilder.rewriteQuery(parseInnerQueryBuilder.get(), queryShardContext);
+            queryBuilder.toFilter(queryShardContext);
```
does this mean that we allow empty filters for aliases? If so can we change that?

javanna replied:
sorry, I guess we will change it for 6.x but keep it for 5.x

javanna commented on core/src/main/java/org/elasticsearch/index/IndexService.java:
```diff
@@ -601,8 +602,11 @@ private Query parse(AliasMetaData alias, QueryShardContext shardContext) {
+                if (innerQueryBuilder.isPresent()) {
+                    return shardContext.toFilter(innerQueryBuilder.get()).query();
+                }
+                return null;
```
here as well, would be nice like above not to have to return null ever. I think if we make validation stricter then we can rely on the filter to always be present?

cbuescher replied:
That's the idea for 6.0, when we remove the possibility of having empty query clauses. Here I think we shouldn't break the current behaviour. The javadoc of the calling method (IndexService#aliasFilter) itself states that potentially returns null, which we would break if we decide to throw an error here or return MatchNone/MatchAll instead.

javanna commented on core/src/main/java/org/elasticsearch/index/query/ConstantScoreQueryBuilder.java:
```diff
@@ -87,10 +88,10 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
         XContentParser parser = parseContext.parser();
 
-        QueryBuilder query = null;
+        Optional<QueryBuilder> query = Optional.empty();
```
I think you can leave null here. this init empty value will never be used as we throw exception when queryFound is false. Otherwise there are two places that can cause empty and that may be confusing.

cbuescher replied:
Right, will change that.

javanna commented on core/src/main/java/org/elasticsearch/index/query/DisMaxQueryBuilder.java:
```diff
@@ -169,8 +168,8 @@ public static DisMaxQueryBuilder fromXContent(QueryParseContext parseContext) th
 
-        if (!queriesFound) {
-            throw new ParsingException(parser.getTokenLocation(), "[dis_max] requires 'queries' field");
+        if (!queriesFound || queries.isEmpty()) {
```
doesn't this beak bw comp as opposed to other changes in this PR? I thought we'd have to bubble the empty up in this case as well.

cbuescher replied:
good point, the old behaviour was to return no documents when there are no clauses (via returning null in toQuery for this case). I will change this back here and return MatchNoDocsQuery in toQuery() for this edge case.

javanna replied:
sounds good

javanna commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FilterAggregationBuilder.java:
```diff
@@ -89,18 +83,11 @@ protected XContentBuilder internalXContent(XContentBuilder builder, Params param
-        }
-
+    public static FilterAggregationBuilder parse(String aggregationName, QueryParseContext context) throws IOException {
+        QueryBuilder filter = context.parseInnerQueryBuilder().orElse(new MatchAllQueryBuilder());
```
weren't we throwing exception before? but now we use match_all instead?

cbuescher replied:
Throwing the exception was introduced by 3197f7f7d856883f66b491270648e532952ad73a and the assumption was right when we represend empty clauses with EmptyQueryBuilder.  We have a test in FilterIT (testEmptyFilter()) that checks that empty clauses should be resolved to MatchAll, which breaks when I removed EmptyQueryBuilder. So I think resolving to MatchAll is right here until we remove support for emtpy clauses altogether.

javanna commented on core/src/main/java/org/elasticsearch/search/rescore/QueryRescorerBuilder.java:
```diff
@@ -56,7 +57,7 @@
         QUERY_RESCORE_PARSER.declareObject(InnerBuilder::setQueryBuilder, (p, c) -> {
             try {
-                return c.parseInnerQueryBuilder();
+                return c.parseInnerQueryBuilder().orElse(QueryBuilders.matchAllQuery());
```
are we sure this shouldn't be null instead?

cbuescher replied:
Allowing null here will later cause NPEs in QueryRescoreBuilder#build() during rewrite. We don't check this in QueryRescoreBuilders ctor yet, but I will add those checks. The current (2.x) behaviour for empty clauses as rescore query is to return all documents in the original query with score unchanged (provided the query/rescore weights are 1.0)

javanna replied:
perfect, thanks for checking

javanna commented:
I had another look and left a few questions but this looks very good!

cbuescher committed:
Adressing review comments

cbuescher commented:
@javanna thanke, I hope I answered and adressed all of your review comments above and in the last commit.

javanna commented on test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java:
```diff
@@ -399,6 +399,7 @@ public void testToQuery() throws IOException {
             QB controlQuery = copyQuery(firstQuery);
             setSearchContext(randomTypes, context); // only set search context for toQuery to be more realistic
             Query firstLuceneQuery = rewriteQuery(firstQuery, context).toQuery(context);
+            assertNotNull("toQuery should not return null", firstLuceneQuery);
```
shall we do the same with the second lucene query and other possible queries we have here?

cbuescher replied:
sure, done.

javanna commented on core/src/main/java/org/elasticsearch/index/query/DisMaxQueryBuilder.java:
```diff
@@ -114,11 +115,13 @@ public float tieBreaker() {
     protected void doXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(NAME);
         builder.field(TIE_BREAKER_FIELD.getPreferredName(), tieBreaker);
+        if (queries.isEmpty() == false) {
```
indentation looks off here. btw shall we output no array or an empty array? :) what were we doing before? not sure this change is needed.

cbuescher replied:
Right, that wasn't supposed to be in the commit, my bad. I will delete it.

javanna commented:
I left a few more comments on your newly added commits, but still looks good ;)

cbuescher commented:
@javanna I adressed the last two comments and pushed an update of the last commit.

javanna commented:
LGTM

cbuescher commented:
@javanna thanks, will merge this then.
